executor/src/main/java/com/redhat/service/smartevents/executor/ExecutorService.java

```diff
@@ -55,17 +63,49 @@ public CompletionStage<Void> processEvent(final Message<String> message) {
         return message.ack();
     }
 
-    private CloudEvent wrapToCloudEvent(String event) {
+    private CloudEvent wrapToCloudEvent(String event, Headers headers) {
         try {
-            return CloudEventBuilder.v1()
+            // JsonCloudEventData.wrap requires an empty JSON
+            JsonNode payload = stringEventToJsonOrEmpty(event);
+            CloudEventBuilder cloudEventBuilder = CloudEventBuilder.v1()
                     .withId(UUID.randomUUID().toString())
                     .withSource(URI.create(CLOUD_EVENT_SOURCE))
                     .withType(String.format("%sSource", executor.getProcessor().getDefinition().getRequestedSource().getType()))
-                    .withData(JsonCloudEventData.wrap(mapper.readTree(event)))
-                    .build();
+                    .withData(JsonCloudEventData.wrap(payload));
+
+            Map<String, Object> extensions = wrapHeadersToExtensionsMap(headers);
+            for (Map.Entry<String, Object> kv : extensions.entrySet()) {
+                // We currently support only String extensions
+                cloudEventBuilder.withExtension(kv.getKey(), kv.getValue().toString());
+            }
+            return cloudEventBuilder.build();
         } catch (JsonProcessingException e2) {
             LOG.error("JsonProcessingException when generating CloudEvent for '{}'", event, e2);
             throw new CloudEventDeserializationException("Failed to generate event map");
         }
     }
+
+    public static Map<String, Object> wrapHeadersToExtensionsMap(Headers kafkaHeaders) {
+        Map<String, Object> resultMap = new HashMap<>();
+        if (kafkaHeaders == null) {
+            return resultMap;
+        }
+
+        for (Header kh : kafkaHeaders) {
+            String cloudEventsExtensionName = adjustExtensionName(kh.key());
+            String headerValue = new String(kh.value(), StandardCharsets.UTF_8);
+            resultMap.put(cloudEventsExtensionName, headerValue);
```

Comment:
There's potential for a problem here if two extension names simplify to the same value.

`my-extension` and `my extension` and `MyExtension` will lead to the others being silently overwritten.

Possibly a difficult scenario to debug. IDK if you'd want to log something or throw an exception?

Reply:
TBH I don't think we're going to ever have the problem, but I think we'll discover it in the future :)